### PR TITLE
[pytorch] Add HashStore to c10d

### PR DESCRIPTION
**Pull request by jjlilley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28921 [pytorch] Add HashStore to c10d**

This implementation is quite similar to the HashStore in gloo -
an ephemeral in-process store with a lock and unordered_map<>.

There are a few tweaks/differences based on c10d vs gloo:
  - c10d expects add/check methods
  - c10d get() use cases expect to wait up to super::timeout_ if the value isn't present
  - c10d set() isn't expected to throw if the value is present.
  - c10d uses uint8_t vs char

It's potentially a better choice for some cases than FileStore when we
don't need cross-process access, or care about the backing file.

Differential Revision: [D18233713](https://our.internmc.facebook.com/intern/diff/D18233713/)

### CODE REVIEW

File: torch/lib/c10d/HashStore.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
+#include <c10d/HashStore.hpp>
+
+#include <stdint.h>
+#include <unistd.h>
+
+#include <errno.h>
```
Comment (pietern) [inline torch/lib/c10d/HashStore.cpp]: Pedantic: can you group C stdlib headers and C++ stdlib headers separately?

File: torch/lib/c10d/HashStore.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  std::unique_lock<std::mutex> lock(m_);
+  auto it = map_.find(key);
+  if (it != map_.end()) {
+    return it->second;
+  }
+  // Not found? Wait up to any Store timeout_.
+  lock.unlock();
+  wait({key}, timeout_);
+  lock.lock();
+  return map_[key];
```
Comment (pietern) [inline torch/lib/c10d/HashStore.cpp]: `cv_.wait_for`? The `pred` can be simpler than the one used in `wait`.

**Review by pietern — APPROVED** [review]
Thanks, @jjlilley!

Two minor nits. Feel free to merge after addressing.

File: torch/lib/c10d/HashStore.cpp (hunk 2 above)
Comment (jjlilley) [reply torch/lib/c10d/HashStore.cpp]: sure, done.

File: torch/lib/c10d/HashStore.cpp (hunk 1 above)
Comment (jjlilley) [reply torch/lib/c10d/HashStore.cpp]: sure, done :)

**Comment by jjlilley** [comment]
Thanks for looking!

**Pull request closed without merging**